aten/src/ATen/native/TensorConversions.cpp

```diff
@@ -489,10 +516,92 @@ Tensor dense_to_sparse_csc(const Tensor& self) {
   return self.to_sparse().to_sparse_csc();
 }
 
+Tensor _tile_tensor(const Tensor& self, IntArrayRef blocksize) {
+  // This code turns a matrix into a sequence of blocks
+  //
+  // Given matrix
+  //
+  //  1  2  3  4
+  //  5  6  7  8
+  //  9 10 11 12
+  // 14 15 16 17
+  //
+  // _tile_tensor(matrix, {2, 2}) will yield the following 2 by 2 blocks
+  //
+  //  1  2 |  3  4 |  9 10 | 11 12
+  //  5  6 |  7  8 | 14 15 | 16 17
+  //
+  //  via a 4D Tensor of shape (2, 2, 2, 2)
+  //
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(blocksize[0] > 0);
+  TORCH_INTERNAL_ASSERT_DEBUG_ONLY(blocksize[1] > 0);
+  auto block_size_0 = self.size(0) / blocksize[0];
+  auto block_size_1 = self.size(1) / blocksize[1];
+  return self.reshape({block_size_0, blocksize[0], block_size_1, blocksize[1]})
+      .transpose(1, 2)
+      .contiguous();
+}
+
+std::pair<Tensor, Tensor> _not_zero_mask_to_col_row_indices(
+    Tensor not_zero_mask) {
+  auto col_indices = at::native::arange(not_zero_mask.size(-1))
```

Comment:
`masked_select` requires that the input and mask use the same device. Hence:
```suggestion
  auto col_indices = at::native::arange(not_zero_mask.size(-1)).to(not_zero_mask.device())
```
(or specify device argument in `arange`)